extensions/amp-consent/0.1/consent-policy-manager.js

```diff
@@ -26,27 +33,147 @@ export const CONSENT_POLICY_STATE = {
 
-    this.policyInstances_ = {};
+    /** @private {!Promise} */
+    this.ConsentStateManagerPromise_ =
+        getServicePromiseForDoc(this.ampdoc_, CONSENT_STATE_MANAGER);
   }
 
   /**
    * Register the policy instance
-   * @param {string} unusedPolicyId
+   * @param {string} policyId
+   * @param {!JsonObject} config
+   */
+  registerConsentPolicyInstance(policyId, config) {
+    dev().assert(!this.instances_[policyId],
+        `${TAG}: instance already registered`);
+
+    const itemsToWait = Object.keys(config['itemsToWait'] || {});
+
+    const instance = new ConsentPolicyInstance(itemsToWait);
+
+    this.instances_[policyId] = instance;
+
+    if (this.policyInstancePromiseResolver_[policyId]) {
+      this.policyInstancePromiseResolver_[policyId]();
+      this.policyInstancePromiseResolver_[policyId] = null;
+      this.policyInstancePromise_[policyId] = null;
+    }
+
+    this.ConsentStateManagerPromise_.then(manager => {
+      for (let i = 0; i < itemsToWait.length; i++) {
+        const consentId = itemsToWait[i];
+        manager.whenConsentReady(consentId).then(() => {
+          manager.onConsentStateChange(consentId, state => {
+            instance.consentStateChangeHandler(consentId, state);
+          });
+        });
+      }
+    });
+  }
+
+
+  /**
+   * Used to wait for policy to resolve;
+   * @param {string} policyId
+   * @return {!Promise}
+   */
+  whenPolicyResolved(policyId) {
+    return this.whenPolicyInstanceReady_(policyId).then(() => {
+      return this.instances_[policyId].getReadyPromise();
+    });
+  }
+
+  /**
+   * Wait for policy instance to be ready.
+   * @param {string} policyId
+   * @return {!Promise}
    */
-  registerConsentPolicyInstance(unusedPolicyId) {
+  whenPolicyInstanceReady_(policyId) {
+    if (this.instances_[policyId]) {
+      return Promise.resolve();
+    }
+    if (!this.policyInstancePromise_[policyId]) {
+      this.policyInstancePromise_[policyId] = new Promise(resolve => {
+        this.policyInstancePromiseResolver_[policyId] = resolve;
+      });
+    }
+    return /** @type {!Promise} */ (this.policyInstancePromise_[policyId]);
+  }
+}
+
+export class ConsentPolicyInstance {
+  constructor(pendingItems) {
+    /** @private {!Array<string>} */
+    this.pendingItems_ = pendingItems;
 
+    /** @private {?function()} */
+    this.readyPromiseResolver_ = null;
+
+    /** @private {!Promise} */
+    this.readyPromise_ = new Promise(resolve => {
+      this.readyPromiseResolver_ = resolve;
+    });
   }
 
   /**
-   * Returns a promise with a consent states value
-   * @param {string} unusedPolicyId
-   * @return {Promise<CONSENT_POLICY_STATE>}
+   * consent instance state change handler
+   * @param {string} consentId
+   * @param {CONSENT_ITEM_STATE} state
    */
-  onPolicyInstanceResolved(unusedPolicyId) {
+  consentStateChangeHandler(consentId, state) {
+    if (state == CONSENT_ITEM_STATE.GRANTED) {
+      const index = this.pendingItems_.indexOf(consentId);
```

Comment:
Could there ever be enough `pendingItem` that array performance becomes a problem? I wish we could use `Set` with closure compiler :(

Reply:
Good point. I added a comment to be aware of the array performance problem. I can always switch to `map` later.

Reply:
the problem with map in cases like this is that checking "length" (keys) is linear time. @kristoferbaxter did that `Set` polyfill ever happen?

Reply:
btw I understand this is a bikeshed for now :) just curious because I wanted to use set a few times now

Reply:
That's not a problem. I only need to access the item state in O(1) time, then I can have another `grantedItemCount` value. 
I wouldn't worry too much. Two consents UIs can ruin user experiences already. That's way before we need to worry about performance.